### PR TITLE
API-1065: Create ConnectionContext service

### PR DESCRIPTION
**Pull request by pierrejolly**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/ConnectionContext.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        $this->username = $username;
+    }
+
+    public function getConnection(): Connection
+    {
+        if (null !== $this->connection) {
+            return $this->connection;
+        }
+
+        $connectionCode = $this->selectConnectionCodeByClientIdQuery->execute($this->clientId);
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/ConnectionContext.php:69]: We need to check if the clientId is not null here **and** if `selectConnectionCodeByClientIdQuery` doesn't return null.
(We should probably fix `DbalSelectConnectionCodeByClientIdQuery` too, that cast a `(string)` on a potentially false value)

It would be ideal if `selectConnectionCodeByClientIdQuery->execute` return null when no connection is found
and that `ConnectionContext->getConnection` return null too.
In case you call this service on a non-api request.

`isCollectable` and `areCredentialsValidCombination` should also handle the case when the service is not initialized and there is no Connection, probably by throwing Exceptions this time.

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/ConnectionContextSpec.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+use Akeneo\Connectivity\Connection\Domain\WrongCredentialsConnection\Persistence\Query\SelectConnectionCodeByClientIdQuery;
+use Akeneo\Connectivity\Connection\Infrastructure\ConnectionContext;
+use PhpSpec\ObjectBehavior;
+
+/**
+ * @author Pierre Jolly <pierre.jolly@akeneo.com>
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class ConnectionContextSpec extends ObjectBehavior
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/ConnectionContextSpec.php:18]: You are missing a lot of tests here

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/EventSubscriber/ConnectionContextEventSubscriberSpec.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+use Akeneo\Connectivity\Connection\Infrastructure\ConnectionContext;
+use Akeneo\Tool\Bundle\ApiBundle\EventSubscriber\ApiAuthenticationEvent;
+use PhpSpec\ObjectBehavior;
+
+/**
+ * @author Pierre Jolly <pierre.jolly@akeneo.com>
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class ConnectionContextEventSubscriberSpec extends ObjectBehavior
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/EventSubscriber/ConnectionContextEventSubscriberSpec.php:16]: Missing tests too

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/ConnectionContext.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+use Akeneo\Connectivity\Connection\Domain\Settings\Persistence\Repository\ConnectionRepository;
+use Akeneo\Connectivity\Connection\Domain\WrongCredentialsConnection\Persistence\Query\AreCredentialsValidCombinationQuery;
+use Akeneo\Connectivity\Connection\Domain\WrongCredentialsConnection\Persistence\Query\SelectConnectionCodeByClientIdQuery;
+
+/**
+ * @author Pierre Jolly <pierre.jolly@akeneo.com>
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class ConnectionContext
```
[the remark this answers is not in the document]
Comment (pierrejolly) [reply src/Akeneo/Connectivity/Connection/back/Infrastructure/ConnectionContext.php:17]: I can't mock this service in my test if I defined the class as final :/

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Dbal/Query/DbalSelectConnectionCodeByClientIdQuery.php [hunk 5]
```diff
@@ -17,14 +17,22 @@ public function __construct(Connection $dbalConnection)
...
 FROM akeneo_connectivity_connection
 WHERE client_id = :client_id
 SQL;
 
-        return (string) $this->dbalConnection->executeQuery($sqlQuery, ['client_id' => $clientId])->fetchColumn();
+        $connectionCode = $this->dbalConnection
+            ->executeQuery($sqlQuery, ['client_id' => $clientId])
+            ->fetchColumn();
+
+        if (null === $connectionCode) {
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Dbal/Query/DbalSelectConnectionCodeByClientIdQuery.php:32]: `fetchColumn` return false if there is not result.

Also, try to use the early return syntax to avoid branching.
```php
if (false === $connectionCode) {
    return null;
}
```

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/EventSubscriber/ConnectionContextEventSubscriberSpec.php [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    ): void {
+        $this->beConstructedWith($connectionContext);
+    }
+
+    public function it_provides_subscribed_events(): void
+    {
+        $this->getSubscribedEvents()->shouldReturn([ApiAuthenticationEvent::class => ['initializeConnectionContext', 1000]]);
+    }
+
+    public function it_initialize_connection_context($connectionContext): void
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/EventSubscriber/ConnectionContextEventSubscriberSpec.php:29]:
```suggestion
    public function it_initializes_connection_context($connectionContext): void
```

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/ConnectionContext.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        $connectionCode = $this->selectConnectionCodeByClientIdQuery->execute($this->clientId);
+
+        if (null === $connectionCode) {
+            return null;
+        }
+
+        return $this->connection = $this->connectionRepository->findOneByCode($connectionCode);
+    }
+
+    public function isCollectable(): bool
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/ConnectionContext.php:81]: Add some phpdoc for the exception `@throw` and I think that using a sub type like `LogicException` could be better.

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/ConnectionContextSpec.php [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        $this->setClientId('client_id');
+        $this->setUsername('username');
+
+        $areCredentialsValidCombinationQuery->execute('client_id', 'username')->willReturn(true);
+        $areCredentialsValidCombinationQuery->execute('client_id', 'username')->shouldBeCalled();
+
+        $this->areCredentialsValidCombination()->shouldReturn(true);
+    }
+
+    public function it_throws_an_exception_during_is_collectable(): void
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/ConnectionContextSpec.php:118]: I think that you should also add a test for when the client_id is not defined and GetConnection return null
